File: chapter_4_trees_graphs/python/4.11-random_node.py
import random

TreeNode = __import__('tree').TreeNode
BinarySearchTree = __import__('binary_search_tree').BinarySearchTree


class BinarySearchTree4_11(BinarySearchTree):
    def insert_in_order(self, value):
        """
        Inserts a node into correct place in Binary Search Tree
        """
        if not self.root:
            self.root = TreeNode4_11(value)
        else:
            self.root.insert_in_order(value)
        self.size += 1

    def get_random_node(self):
        """
        Retrieves a random node in Binary Search Tree
        """
        index = random.randint(0, self.size - 1)
        return self.root.get_ith_node(index)


class TreeNode4_11(TreeNode):
    def __init__(self, value):
        """
        TreeNode constructor with size attribute
        """
        super().__init__(value)
        self.size = 1

    def insert_in_order(self, value):
        """
        Inserts a node into correct place in Binary Search Tree
        """
        if value < self.value:
            if not self.left:
                self.left = TreeNode4_11(value)
            else:
                self.left.insert_in_order(value)
        else:
            if not self.right:
                self.right = TreeNode4_11(value)
            else:
                self.right.insert_in_order(value)
        self.size += 1

    def get_ith_node(self, index):
        """
        Retrieves the node at guven index
        """
        left_side = 0 if self.left == None else self.left.size
        if index < left_side:
            return self.left.get_ith_node(index)
        elif index == left_side:
            return self
        else:
            return self.right.get_ith_node(index - (left_side + 1))


bst = BinarySearchTree4_11()
arr = [20, 30, 10, 15, 35, 17, 5, 3, 7]
for i in arr:
    bst.insert_in_order(i)
string = []
for i in range(20):
    string.append(str(bst.get_random_node().value))
print(" ".join(string))
# 10 15 7 15 30 3 15 15 5 15 10 7 35 20 7 5 35 7 5 3

File: chapter_4_trees_graphs/js/4.11-randomNode.js
const { TreeNode } = require("./tree")
const { BinarySearchTree } = require("./binarySearchTree")

class TreeNode4_11 extends TreeNode {
  constructor(value) {
    /* TreeNode constructor with size attribute */
    super(value)
    this.size = 1
  }

  insertInOrder(value) {
    /* Inserts a node into correct place in Binary Search Tree */
    if (value < this.value) {
      if (!this.left) this.left = new TreeNode4_11(value)
      else this.left.insertInOrder(value)
    } else {
      if (!this.right) this.right = new TreeNode4_11(value)
      else this.right.insertInOrder(value)
    }
    this.size++
  }

  getIthNode(index) {
    /* Retrieves the node at guven index */
    const leftSide = this.left == null ? 0 : this.left.size
    if (index < leftSide) return this.left.getIthNode(index)
    else if (index == leftSide) return this
    else return this.right.getIthNode(index - (leftSide + 1))
  }
}

class BinarySearchTree4_11 extends BinarySearchTree {
  /* Inserts a node into correct place in Binary Search Tree */
  insertInOrder(value) {
    if (!this.root) this.root = new TreeNode4_11(value)
    else this.root.insertInOrder(value)
    this.size++
  }

  /* Retrieves a random node in Binary Search Tree */
  getRandomNode() {
    if (!this.root) return null
    /* Get random index */
    const index = Math.floor(Math.random() * this.size)
    return this.root.getIthNode(index)
  }
}

const bst = new BinarySearchTree4_11()
const arr = [20, 30, 10, 15, 35, 17, 5, 3, 7]
arr.forEach(e => bst.push(e))
const range = Array(10)
let string = ""
for (let i in range) string += bst.getRandomNode().value + " "
console.log(string)
/* 5 10 35 30 7 20 30 17 10 10 */
